ingestion-sink/src/main/java/com/mozilla/telemetry/ingestion/io/BigQuery.java

```diff
@@ -0,0 +1,130 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package com.mozilla.telemetry.ingestion.io;
+
+import com.google.cloud.bigquery.BigQueryError;
+import com.google.cloud.bigquery.InsertAllRequest;
+import com.google.cloud.bigquery.InsertAllResponse;
+import com.google.cloud.bigquery.TableId;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Function;
+
+public class BigQuery {
+
+  public static class Write implements Function<Write.TableRow, CompletableFuture<Write.TableRow>> {
+
+    private final com.google.cloud.bigquery.BigQuery bigquery;
+    private final ConcurrentMap<TableId, Batch> batches = new ConcurrentHashMap<>();
+    private final int maxBytes = 10_000_000; // HTTP request size limit: 10 MB
+    private final int maxMessages = 10_000; // Maximum rows per request: 10,000
+    private final long maxDelayMillis = 1000; // Default to 1 second
+
+    public Write(com.google.cloud.bigquery.BigQuery bigquery) {
+      this.bigquery = bigquery;
+    }
+
+    @Override
+    public CompletableFuture<TableRow> apply(TableRow row) {
+      CompletableFuture<CompletableFuture<TableRow>> future = new CompletableFuture<>();
```

Comment:
AFAICT, the outer future here is just a wrapper that allows us to modify a value inside the lambda below. This is a bit difficult to parse due to the nested futures, so perhaps using an AtomicReference would be better:

```
      AtomicReference<CompletableFuture<TableRow>> futureRef = new AtomicReference<>();
      batches.compute(row.tableId, (tableId, batch) -> {
        if (batch != null) {
          batch.add(row).ifPresent(futureRef::set);
        }
        if (futureRef.get() == null) {
          batch = new Batch(bigquery, row.tableId);
          batch.add(row).ifPresent(futureRef::set);
        }
        return batch;
      });
      return Optional.ofNullable(futureRef.get())
          .orElseThrow(() -> new IllegalArgumentException("Empty batch rejected message"));
```